File listing UI adjustments: sorting, filtering

nl0 opened the pull request:
# Description


# TODO

<!-- Use <del></del> for items that are not required for this PR -->

- [ ] Documentation
    - [ ] JavaScript: basic explanation and screenshot of new features (probably later in a clean PR)
- [x] [Changelog](CHANGELOG.md) entry

fiskus commented on catalog/app/containers/Bucket/Listing.tsx:
```diff
@@ -0,0 +1,850 @@
+  }),
+)
+
+type WrappedAutosizeInputProps = { className?: string } & React.ComponentProps<
```
You can ignore `ref` since it causes error:

```
interface WrappedAutosizeInputProps extends Omit<AutosizeInputProps, 'ref'> {
  className?: string
}
```

nl0 replied:
thx

fiskus commented on catalog/app/containers/Bucket/Listing.tsx:
```diff
@@ -0,0 +1,850 @@
+  const [prefixValue, setPrefixValue] = React.useState(prefix)
+
+  const blur = React.useCallback(() => {
+    if (inputRef.current && inputRef.current.blur) inputRef.current.blur()
```
I think we can adopt optional chaining:

```js
    inputRef?.current?.blur?.()
```

fiskus commented:
I didn't approve because it's WIP, but I heard you said it's ready to merge

nl0 commented:
> I didn't approve because it's WIP, but I heard you said it's ready to merge

it's ready now, i've cleaned it up a little, extracted some unnecessary changes into separate PRs

fiskus commented on catalog/app/containers/Bucket/Listing.tsx:
```diff
@@ -0,0 +1,850 @@
+  const [prefixValue, setPrefixValue] = React.useState(prefix)
+
+  const blur = React.useCallback(() => {
+    inputRef.current?.blur?.()
```
Just checked the resulting code from `./build` directory, it already contains a lot of chaining. But, it seems that chaining operators are transpiled into

![Screenshot from 2021-03-17 15-16-09](https://user-images.githubusercontent.com/533229/111466037-d1ed3e00-8733-11eb-8bfa-8acd22b7f115.png)

nl0 replied:
that's bc we're down-leveling to ES5

fiskus reviewed (approved):
The special directory `".."` should be always on top, it's a common pattern for file managers. But for this listing directory `".."` is moved to the bottom when sorting files, and it disappears when sorting by name directories only.
Also, there is strange behavior for switching asc/desc sorting: when "sort descending" is selecting, I'm clicking on "arrow" and expecting switching to "sort ascending" but sort is resetting

nl0 commented:
@fiskus 

> when "sort descending" is selecting, I'm clicking on "arrow" and expecting switching to "sort ascending" but sort is resetting

that's fine i guess - it just cycles asc/desc/none

> The special directory ".." should be always on top, it's a common pattern for file managers. But for this listing directory ".." is moved to the bottom when sorting files, and it disappears when sorting by name directories only.

ok, i've implemented the desired behavior via some strange workaround, pls have a look

fiskus commented on catalog/app/containers/Bucket/Listing.tsx:
```diff
@@ -614,6 +614,38 @@ function FilteredOverlay() {
+  const va = getValue(a)
+  const vb = getValue(b)
+  // eslint-disable-next-line no-nested-ternary
+  return va < vb ? -1 : va > vb ? 1 : 0
```
Looks almost like Perl
